File: jambeth/jambeth-cache/src/main/java/de/osthus/ambeth/cache/rootcachevalue/DefaultRootCacheValueFactoryDelegate.java
package de.osthus.ambeth.cache.rootcachevalue;

import net.sf.cglib.reflect.FastClass;
import net.sf.cglib.reflect.FastConstructor;
import de.osthus.ambeth.exception.RuntimeExceptionUtil;
import de.osthus.ambeth.merge.model.IEntityMetaData;

public class DefaultRootCacheValueFactoryDelegate extends RootCacheValueFactoryDelegate
{
	protected final FastConstructor constructor;

	public DefaultRootCacheValueFactoryDelegate()
	{
		FastClass fastClass = FastClass.create(Thread.currentThread().getContextClassLoader(), DefaultRootCacheValue.class);
		try
		{
			constructor = fastClass.getConstructor(DefaultRootCacheValue.class.getConstructor(IEntityMetaData.class));
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
	}

	@Override
	public RootCacheValue createRootCacheValue(IEntityMetaData metaData)
	{
		try
		{
			return (RootCacheValue) constructor.newInstance(new Object[] { metaData });
		}
		catch (Throwable e)
		{
			throw RuntimeExceptionUtil.mask(e);
		}
	}
}

File: ambeth/Ambeth.Cache/ambeth/cache/rootcachevalue/DefaultRootCacheValueFactoryDelegate.cs
using De.Osthus.Ambeth.Merge.Model;
using System;
using System.Reflection;

namespace De.Osthus.Ambeth.Cache.Rootcachevalue
{
    public class DefaultRootCacheValueFactoryDelegate : RootCacheValueFactoryDelegate
    {
        protected readonly ConstructorInfo constructor;

        public DefaultRootCacheValueFactoryDelegate()
        {
            ConstructorInfo ci = typeof(DefaultRootCacheValue).GetConstructor(new Type[] { typeof(IEntityMetaData) });
            constructor = typeof(DefaultRootCacheValue).GetConstructor(new Type[] { typeof(IEntityMetaData) });
        }

        public override RootCacheValue CreateRootCacheValue(IEntityMetaData metaData)
        {
            return (RootCacheValue)constructor.Invoke(new Object[] { metaData });
        }
    }
}